add support for PWM balanced mode

youngkin opened the pull request:
Addresses issue #78. It was tested on a Raspberry Pi 3B+ (the only model I have) using [freqtest.go](https://github.com/youngkin/gpio/blob/main/pwmdemo/pwmexplorer/apps/freqtest.go). It can be used to test both balanced and markspace modes. Here's the command line for markspace (assuming the led is wired to GPIO/BCM pin 18):

```
sudo /usr/local/go/bin/go run freqtest.go -pwmType="hardware" -freq="4689" -range="1000" -pulsewidth="10" -pin="18" -pwmmode=1
```

With these settings the LED should blink about 4 times/second (~4Hz). Use this command for balanced mode:

```
sudo /usr/local/go/bin/go run freqtest.go -pwmType="hardware" -freq="4689" -range="1000" -pulsewidth="10" -pin="18" -pwmmode=0
```

WIth these settings (pwmmode=0) the LED should not be blinking (perhaps a minor flicker).

I ran `go fmt` on the file (my editor does this on save). This resulted in some additional minor changes (e.g., whitespace).

youngkin committed:
add support for PWM balanced mode

drahoslove commented on rpio.go:
```diff
@@ -628,11 +639,22 @@ func SetFreq(pin Pin, freq int) {
+// PWM algorithm to use. The constants Balanced or Markspace should be used as the
+// value. See 'SetDutyCycle(pin, dutyLen, cycleLen)' above for more information
+// regarding how to use 'SetDutyCycleMsen()'
+func SetDutyCycleMsen(pin Pin, dutyLen, cycleLen, msen uint32) {
```
I'm not sure about the name of the function.
How about `SetDutyCycleWithMode(pin Pin, dutyLen, cycleLen, mode uint32)` ?

youngkin replied:
I agree, much less cryptic

drahoslove replied:
Well, I'm not really sure what is the better solution.

If you break it into two functions, would you still set the `msen` bit to the default value (1) when setting the DutyCycle, or would you leave the bit intact?

If you would preserve the original behavior of the existing function, then I think I'm probably liking more the idea to implement the second function just for setting the mode.

youngkin replied:
My last comment was more of an editorial than a suggestion to change anything. I do not want to break up the functions.

I'm not sure from your response what you think of my proposal with changing the name to `setDutyCycleWithPWMMode()`.  Are you OK with this or would you like to keep the name you proposed, `SetDutyCycleWithMode()`>

drahoslove replied:
I don't mind either. But please use `Pwm` as if it was a word, not `PWM` (we already have `Pwm` const and  `Pwm()` method, and `SpiStart` instead of `SPIStart`, etc...)
It is the Go convention.

youngkin replied:
>  don't mind either. But please use Pwm as if it was a word, not PWM (we already have Pwm const and Pwm() method, and SpiStart instead of SPIStart, etc...)
It is the Go convention.

Regarding capitalization of acronyms (Google calls them initialisiims), the official [Go code review guidelines](https://github.com/golang/go/wiki/CodeReviewComments#initialisms) call for acronyms to have consistent case. For example, `ServeHTTP` or `xmlHTTPRequest` (the public name would be XMLHTTPRequest). This is why I proposed the name `setDutyCycleWithPWMMode` vs. `setDutyCycleWithPwmMode`.

That said, I am definitely **NOT** suggesting that variable names in this library be changed. In fact, you were quite right to point out that consistency is more important than an arbitrary naming standard. I did notice the use in public function comments of the pattern `FunctionName: description...`. `go vet` will call this out (and did) as non-standard and point out the pattern `FunctionName description...` should be used. But, as in variable naming I chose to be consistent with the current pattern because it was more important to be consistent, not to mention the amount of busy work it would take to make it compliant.

My intent here is not to be annoying (although maybe I am ;) ), but to point out the actual Go conventions for naming variables as well as to provide a link to the [Go code review guidelines](https://github.com/golang/go/wiki/CodeReviewComments). Of course individual organizations and projects are free to set their own standards, but most projects will go with the official Go standards.

drahoslove commented on rpio.go:
```diff
@@ -646,11 +668,12 @@ func SetDutyCycle(pin Pin, dutyLen, cycleLen uint32) {
 	const ctlMask = 255 // ctl setting has 8 bits for each channel
 	const pwen = 1 << 0 // enable pwm
-	const msen = 1 << 7 // use M/S transition instead of pwm algorithm
+	msen = msen << 7
```
The argument should either be bool,
```
if mode == Markspace {
  msen = 1 << 7
}
```
 or you should mask just the single bit
 ```
 msen  = (mode << 7) & (1 << 7)
 ```
  
 Otherwise passing an arbitrary integer > 1 might break something.

youngkin replied:
Good catch, as currently written this is a bug. There is a complication though. If a value other than `Balanced` or `Markspace` is passed it's an error. As currently written the functions don't return an error. To maintain compatibility with existing code, i.e., don't return an error, would require the code to take a default value (e.g., `Markspace`) when provided an invalid value. However, this would be silently masking the error, something I am loath to do.

As a compromise `SetDutyCycle()` can be assumed to always pass a valid value, `Markspace`. `SetDutyCycleMsen()` could be changed to return an error. `SetDutyCycle()` could safely ignore this error. This would maintain backwards compatibility. Any users of `SetDutyCycleMsen()` would be required to handle the error, either by ignoring it (_dangerous_) or explicitly handling it. How do you feel about this approach?

drahoslove replied:
Yes, you are right.
Your proposal sounds good.

Or you can do this:
```
const (
  Balanced = false
  Markspace = true
)

func SetDutyCycleSomethingSomething(pin Pin, dutyLen, cycleLen, mode uint32) {
...
const msen = 0
if mode == Markspace {
  msen = 1 << 7
}
...
}
```
That way it would not be possible to enter an invalid value and you don't have to handle the error.

Your solution is also ok though.

youngkin replied:
I keep forgetting this is Go, not C :) I'm going back and forth between the 2 right now.

So, I like your second proposal, changing `Balanced/Markspace` to a `bool` and then have the method signature take a `bool` for the `mode` parameter. I've got to be more aware of which language I'm writing in :)

drahoslove reviewed (commented):
This is a nice idea, thank you.

Could you address my comments before I merge it?

youngkin commented on rpio.go:
```diff
@@ -628,11 +639,22 @@ func SetFreq(pin Pin, freq int) {
 //   channel 1 (pwm0) for pins 12, 18, 40
 //   channel 2 (pwm1) for pins 13, 19, 41, 45.
 func SetDutyCycle(pin Pin, dutyLen, cycleLen uint32) {
+	SetDutyCycleMsen(pin, dutyLen, cycleLen, 1) // Default PWM algorithm is Mark/Space, msen = 1
+
```
Absolutely, good point

youngkin committed:
address code review comments

youngkin committed:
add an example to test PWM balanced mode

youngkin commented:
This is ready for another review round. I added an example program to demonstrate the new capability and added some comments to help explain all the shifting that's going on.

drahoslove commented:
This looks good, thank you for the contribution!
